File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int n = matrix.length;
        int m = matrix[0].length;

        int rowLow = 0, rowHigh = n-1;

        while(rowLow<=rowHigh){
            int midRow = (rowLow + rowHigh)/2;
            int cLow = 0, cHigh = m-1;
            while(cLow<=cHigh){
                int mid = (cLow + cHigh)/2;
                if(matrix[midRow][mid]==target)
                    return true;
                else if(matrix[midRow][mid]>target){
                    cHigh = mid-1;
                }
                else{
                    cLow = mid+1;
                }
            }
            if(cLow == 0){
                rowHigh = midRow-1;
            }
            else{
                rowLow = midRow+1;
            }
        }

        return false;
        
    }
}

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int s =0;
        int cols = matrix[0].size();
        int rows = matrix.size();
        int e = rows*cols-1;
        int mid = s + (e-s)/2;
        int rowIndex, colIndex;
        int element;
        while(s <= e){
            rowIndex = mid /cols;
            colIndex = mid %cols;
            element = matrix[rowIndex][colIndex];
            if(target == element){
                return true;
            }
            else if(target < element){
                e = mid -1;
            }
            else{
                s = mid +1;
            }
            mid = s + (e-s)/2;
        }
        return false;
    }
};
